charmcraft/charm_builder.py

```diff
@@ -224,21 +229,23 @@ def handle_dependencies(self):
 
         # virtualenv with other dependencies (if any)
         if self.requirement_paths:
-            _process_run(["pip3", "--version"])
-
-            venvpath = self.buildpath / VENV_DIRNAME
-            cmd = [
-                "pip3",
-                "install",  # base command
-                "--target={}".format(venvpath),  # put all the resulting files in that specific dir
-            ]
+            staging_venv_dir = self.charmdir / STAGING_VENV_DIRNAME
+            _process_run(["python3", "-m", "venv", str(staging_venv_dir)])
+            pip_cmd = str(staging_venv_dir / "bin" / "pip3")
+
+            _process_run([pip_cmd, "--version"])
+
+            cmd = [pip_cmd, "install", "--no-binary", ":all:"]  # base command
```

Comment:
If requirements changed, how are we reflecting those changes in the staging venv dir?

Reply:
If requirements are added, only the new requirements are installed and old requirements are not reinstalled. If requirements are removed, they'll remain installed on venv and a rebuild is needed to clean them (this is similar to what happens when source files are removed and sources need to be re-pulled).

Reply:
This is the main benefit of using a venv instead of --target, only changes get reflected, that said, we need an integration test for this and we probably want to add -U

Reply:
Adding `--upgrade` to the list of pip command line options.

Reply:
Regarding tests: we have a unit test that ensures pip is called with the expected arguments, and tests that verify if ops is present in the resulting charm package's venv (the latter runs during the snap smoke tests in both managed and destructive modes). Installing only the changes is part of pip's behavior, we could design a test to verify that but then we would be testing pip instead of our code.

Reply:
Great!